File: app/src/main/java/com/mbg/mbgsupport/demo/java/singleton/DemoSingleton5.java
package com.mbg.mbgsupport.demo.java.singleton;

/***
 * java方式静态内部类方式创建单例的demo
 * created by canwen2003
 */
public class DemoSingleton5 {
    private static final String TAG="DemoSingletonThird";
    private DemoSingleton5(){
    }

    public static DemoSingleton5 getInstance(){
        return SingletonHolder.instance;
    }

    private static class SingletonHolder{
        private final static  DemoSingleton5 instance=new DemoSingleton5();
    }

    public void logInfo(String str){
        System.out.println(TAG+":"+str);
    }

}

File: app/src/main/java/com/mbg/mbgsupport/demo/kotlin/singleton/DemoSingleton5.kt
package com.mbg.mbgsupport.demo.kotlin.singleton
/***
 * Kotlin方式静态内部类方式创建单例的demo ---推荐使用方式
 * created by canwen2003
 */
class DemoSingleton5 private constructor(){

    companion object {
        private const val TAG="DemoSingletonThird"

        @JvmStatic
        fun get():DemoSingleton5{
            return SingletonHolder.holder;
        }
    }

    private object SingletonHolder {
        val holder= DemoSingleton5()
    }

    fun logInfo(str: String){
        println("${TAG}:$str")
    }
}